python_modules/libraries/dagster-tableau/dagster_tableau/assets.py

```diff
@@ -0,0 +1,105 @@
+from typing import Optional, Sequence, cast
+
+from dagster import (
+    AssetExecutionContext,
+    AssetsDefinition,
+    AssetSpec,
+    ObserveResult,
+    Output,
+    _check as check,
+    multi_asset,
+)
+
+from dagster_tableau.resources import (
+    BaseTableauWorkspace,
+    TableauCloudWorkspace,
+    TableauServerWorkspace,
+)
+
+
+def build_tableau_executable_assets_definition(
+    resource_key: str,
+    workspace: BaseTableauWorkspace,
```

Comment:
This is passed to get the site name and the resource class, but it feels a bit awkward. We could pass the `resource_class` and `site_name` instead. The `site_name` could also be added to the metadata of the specs.

Resource class is required for now because users can use either `TableauCloudWorkspace` or `TableauServerWorkspace`.

Reply:
> Resource class is required for now because users can use either TableauCloudWorkspace or TableauServerWorkspace.

Are we able to pull out any of the differing logic between the two into a method which is defined on both? So that the body of this fn can look like:

```python
tableau = getattr(context.resources, resource_key))
with tableau.get_client() as client:
    yield from client.refresh_workbooks(refreshable_workbook_ids)
```

To rephrase, can we remove the need for `workspace` or `resource_class` here and fold that logic into some shared interface?

Reply:
Yeah, that makes sense. 

Updated to do `tableau = cast(BaseTableauWorkspace, getattr(context.resources, resource_key))` instead. The workspace parameter was removed.